### ISSUE TITLE
Strange segfaults in Node.js

### ISSUE THREAD

**Issue by Qard** [mafintosh/napi-macros#12]
Not sure how, but using this module seems to result in segfaults from strange places in Node.js 8. It seems to be _very_ inconsistent, coming from places like the V8 garbage collector, sockets, JS Maps, contextify, and other things. I've been struggling to construct a reduced reproduction case, but I've verified that removing the module eliminates the segfaults completely. 🤔 

My guess is that `NAPI_ARGV_BUFFER_CAST` is being used unsafely as it is using [`napi_get_buffer_info`](https://nodejs.org/api/n-api.html#n_api_napi_get_buffer_info) internally, which explicitly warns about buffer lifetimes.

**Comment by mafintosh** [comment]
Do you have an example? Haven’t seen any issues personally

**Comment by Qard** [comment]
Oh, oops. I meant to open this on [event-loop-delay](https://github.com/mafintosh/event-loop-delay), but it applies here too. 😅 

Anyway, I tried using the event-loop-delay module, which is using napi-macros, and using it would result in the process randomly crashing. I tried to make a reduced test case, but wasn't able to reproduce it in isolation. However, removing that one module completely eliminated the segfaults. It went from almost _always_ segfaulting to _never_ segfaulting after removing it.

I'm trying to use event-loop-delay in [this PR to elastic-apm-node](https://github.com/elastic/apm-agent-nodejs/pull/1021/files#diff-f4a2093d1bf303e6e6894f7bcc66d3eeR9) alongside `perf_hooks.monitorEventLoopDelay(...)`, but the test suite just crashes constantly after adding it. I've removed it for the time being and trying to figure out a way to reduce the reproduction case to something actually reasonable. 😓

**Comment by mafintosh** [comment]
Ah makes sense. Most likely gc that’s messing with you. Can you show me exactly where you added it?

**Comment by Qard** [comment]
I used it as a fallback for `perf_hooks.monitorEventLoopDelay()` like this:

```js
let getLoopDelay
try {
  const sampler = require('perf_hooks').monitorEventLoopDelay()
  sampler.enable()
  getLoopDelay = () => {
    const result = sampler.mean || 0
    sampler.reset()
    return result
  }
} catch (err) {
  const eventLoopDelay = require('event-loop-delay')
  let sampler = eventLoopDelay()
  getLoopDelay = () => {
    const result = sampler.delay / sampler.times
    sampler.destroy()
    sampler = eventLoopDelay()
    return result
  }
}
```

The `getLoopDelay()` function is called at some user-configurable interval, defaulting to every 30 seconds. The full PR is here: https://github.com/elastic/apm-agent-nodejs/pull/1021

**Comment by SimonSimCity** [comment]
I don't know how much this helps, but there's another library for tracking the `event loop lag` only using JavaScript: https://github.com/STRML/node-toobusy

The repository is forked from one which also started off as one having C code but it also had problems compiling against more recent node-versions - that's why people forked off and created a version in pure JavaScript.

**Comment by mafintosh** [comment]
Fixed in 1.0.1 - There was a timer bug in C land.
